rust/actyx/ax-aql/src/language/type_check.rs

```diff
@@ -0,0 +1,575 @@
+use super::workflow::{Participant, Workflow, WorkflowStep};
+use crate::{Ident, Ind, Index, Label, NonEmptyVec, Num, Query, SimpleExpr, Type, TypeAtom};
+use std::{collections::BTreeSet, ops::Deref, sync::Arc};
+pub(crate) struct QueryTypeCheckError<'a> {
+    pub(crate) span: pest::Span<'a>,
+    pub(crate) error_type: QueryTypeCheckErrorType,
+}
+
+type ExtraMessage = String;
```

Comment:
This is usually not done in Rust: `type` aliases are used to write down complex types with less verbosity, otherwise newtypes like `struct ExtraMessage(String)` are used to create a separate type with different semantics.

Reply:
With `struct` how can I make it so that `ExtraMessage` can be operated as if it is a string?

Reply:
I want to state that this is a string, but also an "ExtraMessage" for `InvalidIndex` error.
The reason is, `DrillDownResult` cannot be `Result<Type, QueryTypeCheckErrorType::InvalidIndex>` and `Result<Type, String>` isn't clear enough on what the string are

Reply:
In that case I’d just use `Result<Type, QueryTypeCheckErrorType>` as the return type.